Validations added to trim invalid solution paths in solver

anjan-keysight opened the pull request:
@greg-dennis  I have made some initial changes in the solver to validate partial solutions and trim solutions that will definitely fail down those paths. This is for faster solver logic. Do have a look. 

But overall the current solver design approach will have some scaling issues. We tried with 8 port lag (one non lag port); worst case it was taking more than 10 mins. With this fix that has now come down to 5-6 sec. But even with this fix a 12 port lag testbed will take close to 10 mins; and it exponentially increases with lag port size. With multiple DUT and port groups it will compound the problem even more. I think we need to revisit the solver algorithm.

If its ok we can sync up on a call (with @biplamal also)

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -282,7 +338,7 @@ func (s *solver) solve() (*assign, error) {
 			}
 		}
-		port2IntfChan := genCombos(port2Intfs)
+		port2IntfChan := genCombos(port2Intfs, s.checkPort)
```
I'm also wondering if a key problem here is that we're searching through port combinations rather than _link_ combinations. If instead of creating port2Intfs we could create a link2Links maps that takes device, port, and link into consideration. Then there would need to be one smaller added phase to map the unlinked ports. Would that give you at least the same effective pruning that this accomplishes and perhaps more?

anjan-keysight replied:
Yes, the key issue is the port2Intfs (link2Links makes it better). However, I think, the generic issue is generating all solution sets some of which may be invalid. So whether for DUT or PORT if at any iteration, the key -> value selection may lead to invalid solutions lets not proceed down that branch. Also, I think, the PORT selection should be in conjunction with DUT selection (not two separate process); because that will prune a lot of invalid combinations for larger topologies.

anjan-keysight commented on knebind/solver/solver.go:
```diff
@@ -242,6 +254,50 @@ type solver struct {
+
+type check func(interface{}, interface{}, map[interface{}][]interface{}, *[]interface{}, map[interface{}]interface{}, map[interface{}]bool, bool) bool
+
+func (s *solver) checkPort(k interface{}, v interface{}, m map[interface{}][]interface{}, keys *[]interface{}, res map[interface{}]interface{}, used map[interface{}]bool, reserve bool) bool {
```
The fix I have added will be effective only with port group support (PR raised by @biplamal). However I have added a negative test for DUT selection. The issue is similar here; it tries all combinations of DUT (2 Arista X 2 Cisco X 2 Juniper); even if there are no direct links between them.
In port group the issue is, if a lag intf is selected for a non-lag port all subsequent combinations will fail is there are exactly same number of lag ports configured as requested by testbed. Probably I can explain it better over a call.

anjan-keysight commented on knebind/solver/solver.go:
```diff
@@ -242,6 +254,50 @@ type solver struct {
+
+type check func(interface{}, interface{}, map[interface{}][]interface{}, *[]interface{}, map[interface{}]interface{}, map[interface{}]bool, bool) bool
+
+func (s *solver) checkPort(k interface{}, v interface{}, m map[interface{}][]interface{}, keys *[]interface{}, res map[interface{}]interface{}, used map[interface{}]bool, reserve bool) bool {
```
Agree with you on this, but that function type was generic so as to support both DUT => NODE and PORT => INTF selections; so pass everything that would be required to validate and update. If you want we can embed those parameters in a type structure and pass by reference.

anjan-keysight commented:
@greg-dennis I have made new set of changes for faster solutions, based on links and also added support for port grouping. Here are summary of the changes and design approach.
1. For link (with group support) based problem solving I have two types link and lag. For every dev => peer in testbed there is a lag object with link map and ordered (descending) list of lag names between them; the ordering is based on the lag size. This is also generated for topology node => peer mappings.
2. For the generating device => node solutions, I have updated the genCombo / genRecurse at two places. First when getting the list of devices (in the beginning), here the list order is important. Devices are added in the order they appear linked to some peer that is already added. So we start with a device, add all peer, then add all peer of peers and so on. All isolated device are added at the end.
3. The second update is (in genRecurse) that validates an intermediate selection (device => node) based on lag support. Priority is given to largest LAG if satisfied, then next. Regular link (no LAG), match happens last and can be assigned any unused link.
4. Based on device => node combos, we get the corresponding dev link => node link mappings. Again the keys are ordered here in descending order.
5. I have removed generics because of the layered structures, type casting fields over iterations may be costly.
6. I have added tests for grouping and also a scaled topology for performance comparison. With the existing solution it takes more than 10 mins (cutoff time). With this update it takes just over a sec to generate reservation.

greg-dennis commented:
I have received the latest update, thanks, and will begin incorporating it internally

greg-dennis commented:
I've also taken your TestSolveScale test and merged it in as a go Benchmark. Could you rebase to pick up that change?

Once that is done, would it be easy to turn this PR into two PRs: one that trims the solution paths _without_ adding in support for port groups, and a second that adds port group support?

ashutshkumr commented:
@greg-dennis thanks for update here. Anjan is on PTO, so I'll be taking a look at it. Will get back with more updates.

anjan-keysight commented:
@greg-dennis I will update this PR to trim solution paths only and raise a separate for the port group support.

anjan-keysight commented:
@greg-dennis I have updated this PR to address trimming solution paths, group support has been removed. Once merged I will raise the group support PR after resolving with main.

greg-dennis commented:
@anjan-keysight Thanks! We just finished up a team summit, and I'll set to work on this ASAP

greg-dennis commented:
This has been merged. The performance gains on the benchmark test were significant. No performance gain could be measured from arrangeDevKeys, so that function was dropped. There are some other significant renames and restructuring from the submitted PR, so please rebase carefully. Thanks!
